### ISSUE TITLE
Error with shared VPC network after upgrading to 7.3.0

### ISSUE THREAD

**Issue by sebastianneb** [terraform-google-modules/terraform-google-kubernetes-engine#472]
Hallo everyone!

Terraform version: `0.12.23`.

After upgrading from `7.2.0` to `7.3.0` we get an error when applying a GKE cluster in a shared vpc host project environment.

The following code is the minimum I can give you (I deleted some details). The service account was created separately and has all necessary rights (like networkUser on the xpn and the subnet and all other needed rights for logging / monitoring / registry / etc.):

```terraform
module "cluster" {
  source  = "terraform-google-modules/kubernetes-engine/google//modules/private-cluster"
  version = "7.3.0"

  # Default cluster settings
  project_id                    = "serviceProjectA"
  name                          = "app"
  regional                      = true
  region                        = "europe-west4"
  zones                         = ["europe-west4-a", "europe-west4-b"]
  remove_default_node_pool      = true
  service_account               = "app-cluster-nodes@serviceProjectA.iam.gserviceaccount.com"
  grant_registry_access         = false
  create_service_account        = false
  horizontal_pod_autoscaling    = false
  kubernetes_version            = "1.15.9-gke.22"
  enable_private_nodes = true

  # Network settings
  network_project_id    = "shared-vpc-host-project"
  network                       = "xpn"
  subnetwork                 = "app-cluster-subnet"
  ip_range_pods             = "pods-range"
  ip_range_services       = "services-range"
  master_ipv4_cidr_block = "10.0.32.128/28"
}
```

With `7.2.0` this succeeds, with `7.3.0` we get the error in terraform: `Error waiting for creating GKE cluster: Google Compute Engine: Required 'compute.networks.get' permission for 'projects/shared-vpc-host-project/global/networks/xpn'.`

When I output the self_link that was used before [this change](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/commit/8571f61dcd35eb4bd06febea9d14b0ed409b2d0e), the self_link was `https://www.googleapis.com/compute/v1/projects/shared-vpc-host-project/global/networks/xpn`.

So to sum it up: In my opinion this change from not using self_link broke our GKE shared VPC setup.

Happy to provide more details if needed :)

**Comment by zolech** [comment]
I had this problem on 6.0.0 and this post helped me to resolve it https://github.com/terraform-providers/terraform-provider-google/issues/1711#issuecomment-420484808

**Comment by sebastianneb** [comment]
We tried that as well. The serviceAccount that we use to deploy the cluster, as well as the service-account for the GKE, had networkUser (and even when we tried out owner) rights on the xpn and on the specific subnet as well. We were still not able to solve the issue with `7.3.0`

**Comment by zolech** [comment]
Let's clarify. 

Did you set `roles/compute.networkUser` on `somename@container-engine-robot.iam.gserviceaccount.com` in you vpn host project?

Or did you set this role on GKE service account in your gke project.

The reason I am asking is because I did the second thing and it was a mistake. The first one resolved the issue.

**Comment by sebastianneb** [comment]
We (as recommended by i.e. google-project-factory) disabled the default service account of each project, and therefore also in the shared vpc project. 

But to clarify: No, I did not grant `roles/compute.networkUser` to `somename@container-engine-robot.iam.gserviceaccount.com` because we simply do not need it and the service account is nevertheless disabled. Happy to try out if that works but I still think that the reported behavior I posted is a bug.

Do you know why using `self_link` on the network was removed?

**Comment by zolech** [comment]
No I don't know that, I am just a simple users that stumbled uppon this issue.

> We (as recommended by i.e. google-project-factory) disabled the default service account of each project, and therefore also in the shared vpc project.

Ok, but default service account is not being used here.

> But to clarify: No, I did not grant roles/compute.networkUser to somename@container-engine-robot.iam.gserviceaccount.com because we simply do not need it and the service account is nevertheless disabled. Happy to try out if that works but I still think that the reported behavior I posted is a bug.

This is weird to me because this account is needed to provision GKE on shared vpc network.

**Comment by sebastianneb** [comment]
> No I don't know that, I am just a simple users that stumbled uppon this issue.

Ah sorry then, I thought you might know ;) 

> This is weird to me because this account is needed to provision GKE on shared vpc network.

We don't need the default service account since we tell GKE (and all node pools) to use the already created service account, which has all necessary rights. So If am not wrong you do not need the default compute service account if you provide one.

**Comment by zolech** [comment]
I also tried to deploy new pool manualy by google console and also get this problem, not only with terraform. That is why I started to search for permissions problem on my service accounts.

**Comment by morgante** [comment]
Unfortunately I don't think this is a bug and I'm not sure how it was ever working for you actually.

> But to clarify: No, I did not grant roles/compute.networkUser to somename@container-engine-robot.iam.gserviceaccount.com because we simply do not need it and the service account is nevertheless disabled. Happy to try out if that works but I still think that the reported behavior I posted is a bug.

This is required, please add it. To clarify, there are two service accounts involved in the GKE cluster:
- The *node* service account. This is the custom service account you've specified and used (by default, this would be the Compute Engine service account).
- The GKE service account (somename@container-engine-robot.iam.gserviceaccount.com). This account is used by GKE to perform cluster operations and cannot be overridden. It is always needed.

Both service accounts need `roles/compute.networkUser` on the host project. Please add it and verify this fixes your issue.

This doc explains the two accounts in more detail: https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-shared-vpc

**Comment by sebastianneb** [comment]
Sorry for the delayed response. I was reading through some docs and can confirm, that we need those roles that you mentioned @morgante. I checked that and as it turned out, the project-factory terraform module already created these bindings. So I actually had and still have `roles/networkUser` for `service-${projectNumber}@container-engine-robot.iam.gserviceaccount.com` assigned in the host project (as also mentioned on the page you shared).

For some reason, it started working now... (But I am sure, that we had that rights assigned before). Maybe some inconsistency in the terraform state (we had similar issues today with `data`-"resources").

Nevertheless thanks for the help and I will close this issue now :)
